shared/lib/transactions-controller-utils.js

```diff
@@ -34,9 +34,14 @@ export function toPrecisionWithoutTrailingZeros(n, precision) {
     .replace(/(\.[0-9]*[1-9])0*|(\.0*)/u, '$1');
 }
 
-export function calcTokenAmount(value, decimals) {
-  const multiplier = Math.pow(10, Number(decimals || 0));
-  return new BigNumber(String(value)).div(multiplier);
+/**
+ * @param {number|string|BigNumber} value
+ * @param {number} decimals
+ * @returns {BigNumber}
+ */
+export function calcTokenAmount(value, decimals = 0) {
```

Comment:
It seems we have similar logic spread throughout the code, such as:

- `ui/ducks/send/send.js`
- `ui/components/ui/token-input/token-input.component.js`
- `ui/components/app/currency-input/currency-input.js`
- `ui/pages/confirmations/components/confirm/info/typed-sign/permit-simulation/permit-simulation.tsx`
- `ui/pages/confirmations/components/confirm/row/dataTree.tsx`

Should we assume this could also be an issue in those other usages and have them all point to a single utility?

Also, I appreciate this was historic, but there is no `TransactionController` directly in the extension anymore, so this file feels somewhat like a collection of miscellaneous utility functions.

Should we relocate and introduce typings via a new file such as `shared/lib/[math|large-number|decimals].ts`?

Reply:
re: similar logic spread throughout the code
Great callout! I mentioned in this [related PR](https://github.com/MetaMask/metamask-extension/pull/25741#issuecomment-2225556743) that we should reuse this logic there after this is merged (this would address 2 of the bullet points above). I agree it'd be a good idea to update the other code. This is out of this PR. Created a new issue for this to track:
https://github.com/MetaMask/metamask-extension/issues/25891

re: no TransactionController directly in the extension anymore
Another great callout! I also agree here that we should rename/refactor the file accordingly. As this is another refactor and out of scope from this bug fix, I've created a new ticket for this here:
https://github.com/MetaMask/metamask-extension/issues/25892